rust-webvr/src/api/vrexternal/display.rs

```diff
@@ -0,0 +1,265 @@
+        data.left_eye_parameters.field_of_view.right_degrees = l_fov.rightDegrees;
+        data.left_eye_parameters.field_of_view.down_degrees = l_fov.downDegrees;
+        data.left_eye_parameters.field_of_view.left_degrees = l_fov.leftDegrees;
+
+        data.right_eye_parameters.field_of_view.up_degrees = r_fov.upDegrees;
+        data.right_eye_parameters.field_of_view.right_degrees = r_fov.rightDegrees;
+        data.right_eye_parameters.field_of_view.down_degrees = r_fov.downDegrees;
+        data.right_eye_parameters.field_of_view.left_degrees = r_fov.leftDegrees;
+
+        data
+    }
+
+    fn immediate_frame_data(&self, near_z: f64, far_z: f64) -> VRFrameData {
+        let sys = self.shmem.system();
+
+        let mut data = VRFrameData::default();
+
+        data.pose.position = Some(sys.as_ref().sensorState.pose.position);
+        data.pose.orientation = Some(sys.as_ref().sensorState.pose.orientation);
+        data.left_view_matrix = sys.as_ref().sensorState.leftViewMatrix;
+        data.right_view_matrix = sys.as_ref().sensorState.rightViewMatrix;
+
+        let proj = |fov: mozgfx::VRFieldOfView| -> [f32; 16] {
+            use std::f64::consts::PI;
+
+            // FIXME: Note for reviewer: How to get the handedness?
+            let right_handed = true;
+
+            let up_tan = (fov.upDegrees * PI / 180.0).tan();
+            let down_tan = (fov.downDegrees * PI / 180.0).tan();
+            let left_tan = (fov.leftDegrees * PI / 180.0).tan();
+            let right_tan = (fov.rightDegrees * PI / 180.0).tan();
+            let handedness_scale = if right_handed { -1.0 } else { 1.0 };
+            let pxscale = 2.0 / (left_tan + right_tan);
+            let pxoffset = (left_tan - right_tan) * pxscale * 0.5;
+            let pyscale = 2.0 / (up_tan + down_tan);
+            let pyoffset = (up_tan - down_tan) * pyscale * 0.5;
+            let mut m = [0.0f32; 16];
+            m[0 * 4 + 0] = pxscale as f32;
+            m[1 * 4 + 1] = pyscale as f32;
+            m[2 * 4 + 0] = (pxoffset * handedness_scale) as f32;
+            m[2 * 4 + 1] = (-pyoffset * handedness_scale) as f32;
+            m[2 * 4 + 2] = (far_z / (near_z - far_z) * -handedness_scale) as f32;
+            m[2 * 4 + 3] = handedness_scale as f32;
+            m[3 * 4 + 2] = ((far_z * near_z) / (near_z - far_z)) as f32;
+            m
+        };
+
+        let left_fov =
+            sys.as_ref().displayState.mEyeFOV[mozgfx::VRDisplayState_Eye_Eye_Left as usize];
+        let right_fov =
+            sys.as_ref().displayState.mEyeFOV[mozgfx::VRDisplayState_Eye_Eye_Right as usize];
+
+        data.left_projection_matrix = proj(left_fov);
+        data.right_projection_matrix = proj(right_fov);
+
+        data.timestamp = utils::timestamp();
+
+        data
+    }
+
+    fn synced_frame_data(&self, near_z: f64, far_z: f64) -> VRFrameData {
+        // FIXME: Note for reviewer: weren't we supposed to block like with sync_poses?
```

Comment:
blocking in `sync_poses` is enough. Both methods will be called in a row in Servo WebVR. They are separated just because the `near_z` and `far_z` params could change between the calls.